frame/uniswap-v2/src/lib.rs

```diff
@@ -351,6 +390,108 @@ pub mod pallet {
 		}
 	}
 
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			let current_timestamp = T::Time::now();
+			Pools::<T>::iter()
+				.filter(|(pool_id, pool)| T::EnabledPoolTWAP::contains(&(pool.pair, *pool_id)))
+				.fold(Weight::zero(), |weight, (pool_id, _)| {
+					let rate_base = Self::do_get_exchange_rate(pool_id, false);
+					let rate_quote = Self::do_get_exchange_rate(pool_id, true);
+					match (rate_base, rate_quote) {
+						(Ok(rate_base), Ok(rate_quote)) => PriceCumulativesState::<T>::try_mutate(
+							pool_id,
+							|price_cumulatives| -> Result<Weight, DispatchError> {
+								match *price_cumulatives {
+									// Initialization
+									None => {
+										let (price_cumulative_base, price_cumulative_quote) = (
+											compute_initial_price_cumulative::<T::Convert, _>(
+												rate_base,
+											)?,
+											compute_initial_price_cumulative::<T::Convert, _>(
+												rate_quote,
+											)?,
+										);
+										*price_cumulatives = Some(PriceCumulatives {
+											timestamp: current_timestamp,
+											price_cumulative_base,
+											price_cumulative_quote,
+										});
+										TWAPState::<T>::insert(
+											pool_id,
+											TWAP {
+												average_price_base: rate_base,
+												average_price_quote: rate_quote,
+											},
+										);
+										Ok(weight + 1)
+									},
+									// Update
+									Some(PriceCumulatives {
+										timestamp: previous_timestamp,
+										price_cumulative_base: previous_price_cumulative_base,
+										price_cumulative_quote: previous_price_cumulative_quote,
+									}) => {
+										let f = |previous_cumulative_price, exchange_rate| {
+											compute_next_price_cumulative::<T::Convert, _, _>(
+												previous_timestamp,
+												previous_cumulative_price,
+												current_timestamp,
+												exchange_rate,
+											)
+										};
+										let (
+											(elapsed, price_cumulative_base),
+											(_, price_cumulative_quote),
+										) = (
+											f(previous_price_cumulative_base, rate_base)?,
+											f(previous_price_cumulative_quote, rate_quote)?,
+										);
+										if elapsed >= T::TWAPInterval::get() {
```

Comment:
why we are doing other calculations if delta < interval? may be should do delta and only than calculate?